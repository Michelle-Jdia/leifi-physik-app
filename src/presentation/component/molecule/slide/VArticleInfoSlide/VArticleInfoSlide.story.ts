import type { Meta, StoryFn } from '@storybook/vue3';
import { articleInfoSlideProps } from '@/molecule/slide/VArticleInfoSlide/VArticleInfoSlide';
import VArticleInfoSlide from '@/molecule/slide/VArticleInfoSlide/VArticleInfoSlide.vue';

export default {
    component: VArticleInfoSlide,
    argTypes: {
        onOnDelete: {},
        onOnCreate: {},
        onOnFavorite: {},
    },
} as Meta<typeof VArticleInfoSlide>;

const Template: StoryFn<typeof VArticleInfoSlide> = (args) => ({
    components: {
        VArticleInfoSlide,
    },
    setup() {
        return {
            args,
        };
    },
    template: `
        <v-article-info-slide v-bind="args"/>
    `,
});

export const Base = Template.bind({});

export const Topic = Template.bind({});

Base.args = {
    ...articleInfoSlideProps,
};

Topic.args = {
    ...articleInfoSlideProps,
    isTopic: true,
};
