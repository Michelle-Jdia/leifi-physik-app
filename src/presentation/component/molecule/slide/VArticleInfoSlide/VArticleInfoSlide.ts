export const articleInfoSlideProps = {
    id: '1',
    branch: {
        title: 'Mechanik',
        color: 'v-bg-yellow-80',
    },
    title: 'Title',
    subTitle: 'Subtitle',
};
