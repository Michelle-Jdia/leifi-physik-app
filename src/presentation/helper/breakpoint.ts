export const breakpoints = {
    mobile: 576,
    tablet: 768,
};
