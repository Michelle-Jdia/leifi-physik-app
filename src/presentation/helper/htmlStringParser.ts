export const html =
    '<div class="rich-text-main"><div class="process_glossary"><p> </p><p>Zu vielen wichtigen Erkenntnissen über weißes Licht und seine Spektralfarben gelangte Newton mit einfachen Versuchen, die du <a href="/node/7678" target="_blank">hier</a>ausführlicher beschrieben findest.</p><h4 style="clear:both; margin-bottom:1em">Zerlegung von weißem Licht in die Spektralfarben</h4><figure role="group" class="rs_skip caption caption-drupal-entity"><div data-embed-button="media" data-entity-embed-display="view_mode:media.image_50_percent_center" data-entity-type="media" data-entity-uuid="73a22760-ff57-4a31-bac4-bc099cee2e72" class="align-right embedded-entity" data-langcode="de" data-entity-embed-display-settings="[]"><picture class=""><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/fa62eb7a83830f00436cca015aa94f39/576spektralfarben_prismenspektrum_bild.svg" media="(max-width: 570px)" type="image/webp"><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/fa62eb7a83830f00436cca015aa94f39/576spektralfarben_prismenspektrum_bild.svg" media="(max-width: 570px)" type="image/jpeg"><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/fa62eb7a83830f00436cca015aa94f39/992spektralfarben_prismenspektrum_bild.svg" media="(min-width: 571px)" type="image/webp"><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/fa62eb7a83830f00436cca015aa94f39/992spektralfarben_prismenspektrum_bild.svg" media="(min-width: 571px)" type="image/jpeg"><img src="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/fa62eb7a83830f00436cca015aa94f39/1000spektralfarben_prismenspektrum_bild.svg" alt="" /></source></source></source></source></picture><span class="image-copyright font-italic small rs_skip"><i class="far fa-copyright"></i><span class="copyright-text">Joachim Herz Stiftung</span></span></div><figcaption>Abb. 1 Zerlegung von weißem Licht in Spektralfarben mit Hilfe eines Prismas</figcaption></figure><p style="text-align:justify">Wenn du weißes Licht in geeigneter Weise auf ein Prisma fallen lässt, so kannst du auf einem Schirm dahinter beobachten, dass das weiße Licht in verschiedene Farben zerlegt wurde. Auf dem Schirm ist jeweils ein Bereich zu sehen, der rot, orange, gelb, grün, cyan, blau oder violett gefärbt ist. Diese Farben nennt man die <strong>Spektralfarben</strong>.</p><p style="text-align:justify">Die Spektralfarben kannst du auch bei einem Regenbogen finden. Hier ist die Sonne die Quelle für weißes Licht und die Regentropfen sorgen für die Zerlegung.</p><h4 style="clear:both; margin-bottom:1em">Unzerlegbarkeit der Spektralfarben</h4><figure role="group" class="rs_skip caption caption-drupal-entity"><div data-embed-button="media" data-entity-embed-display="view_mode:media.image_50_percent_center" data-entity-type="media" data-entity-uuid="69987120-eeaa-483c-9032-b04e0511d4c0" class="align-right embedded-entity" data-langcode="de" data-entity-embed-display-settings="[]"><picture class=""><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/215ed628eaebf3845102ea5ade417e85/576spektralfarben_unzerlegbarkeit_bild.svg" media="(max-width: 570px)" type="image/webp"><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/215ed628eaebf3845102ea5ade417e85/576spektralfarben_unzerlegbarkeit_bild.svg" media="(max-width: 570px)" type="image/jpeg"><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/215ed628eaebf3845102ea5ade417e85/992spektralfarben_unzerlegbarkeit_bild.svg" media="(min-width: 571px)" type="image/webp"><source srcset="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/215ed628eaebf3845102ea5ade417e85/992spektralfarben_unzerlegbarkeit_bild.svg" media="(min-width: 571px)" type="image/jpeg"><img src="https://www.app-api-dm4xeuy-bwyzh75a7hm5c.eu-2.platformsh.site/sites/default/files/images/215ed628eaebf3845102ea5ade417e85/1000spektralfarben_unzerlegbarkeit_bild.svg" alt="" /></source></source></source></source></picture><span class="image-copyright font-italic small rs_skip"><i class="far fa-copyright"></i><span class="copyright-text">Joachim Herz Stiftung</span></span></div><figcaption>Abb. 2 Unzerlegbarkeit der Spektralfarben</figcaption></figure><p style="text-align:justify">Wenn du nun eine beliebige Spektralfarbe wieder durch ein Prisma fallen lässt, so siehst du auf dem Schirm, dass sich die Farbe des Lichtes nicht wieder geändert hat. Spektralfarben lassen sich nicht weiter zerlegen. Man nennt sie auch reine Farben.</p><h4 style="clear:both; margin-bottom:1em">Nicht sichtbare Strahlung im Spektrum</h4><p style="text-align:justify">Neben den Spektralfarben enthält zum Beispiel das Sonnenlicht noch weitere Anteile, die du nicht mit bloßem Auge sehen kannst - auch dann nicht, wenn du Sonnenlicht durch ein Prisma in seine Spektralfarben zerlegst.</p><p style="text-align:justify">Ein Anteil ist <strong>infrarote Strahlung</strong>(<strong>IR-Strahlung</strong>). Diese Strahlung wäre im Experiment auf dem Schirm (vgl. <strong>Abb. 1</strong>) oberhalb des roten Lichtes zu sehen. Infrarote Strahlung wird zum Beispiel bei Wärmelampen genutzt.</p><p style="text-align:justify">Ein weiterer nicht sichtbarer Teil wird als <strong>ultraviolette Strahlung</strong>(<strong>UV-Strahlung</strong>) bezeichnet. Diese Strahlung wäre im Experiment auf dem Schirm (vgl. <strong>Abb. 1</strong>) unterhalb des blauen Lichtes zu sehen. Ultraviolette Strahlung wird zum Beispiel zum Prüfen von Geldscheinen und zum Härten von Nagellack verwendet. Auch sind Teile der UV-Strahlung für den Sonnenbrand verantwortlich.</p></div></div><div class="clearfix"></div>';

const parser = new DOMParser();

export const doc = parser.parseFromString(html, 'text/html');
