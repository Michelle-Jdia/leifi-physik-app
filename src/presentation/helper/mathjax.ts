/*
 configuration is taken from 'https://www.leifiphysik.de/themes/custom/avonis/js/mathjax-config.js?s0e9g7'
 - instead of the 'tex-mml-chtml' the 'tex-mml-svg' is used displaying the rendered functions as svg not with css & html causing errors by including fonts
 - the config for loader is removed and all the extensions are imported separately, otherwise causing errors
 */

/* eslint-disable */
// This is no longer in use. See public/script/mathjax
// export const getMathjax = async () => {
//     // @ts-ignore
//     await import('/public/script/mathjax/mathjax-config.js');
//
//     // @ts-ignore
//     await import('/public/script/mathjax/tex-mml-chtml.js');
//
//     // @ts-ignore
//     import('/public/script/mathjax/input/tex/extensions/enclose.js');
//
//     // @ts-ignore
//     import('/public/script/mathjax/input/tex/extensions/cancel.js');
//
//     // @ts-ignore
//     import('/public/script/mathjax/input/tex/extensions/color.js');
//
//     // @ts-ignore
//     // import('/public/script/mathjax/input/tex/extensions/colorv2.js');
//
//     // @ts-ignore
//     import('/public/script/mathjax/input/tex/extensions/mhchem.js');
// };
