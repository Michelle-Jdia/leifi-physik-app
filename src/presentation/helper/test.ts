import type { ComponentMountingOptions } from '@vue/test-utils';
import type { Component } from 'vue';
import { config, mount } from '@vue/test-utils';
import { expect, test } from 'vitest';

beforeAll(() => {
    config.global.renderStubDefaultSlot = true;
});

afterAll(() => {
    config.global.renderStubDefaultSlot = false;
});

export const testSnapshot = (
    component: Component,
    options?: ComponentMountingOptions<object>,
): void => {
    test('snapshot', () => {
        const wrapper = mount(component, {
            shallow: true,
            ...options,
        });

        expect(wrapper.element).toMatchSnapshot();
    });
};
