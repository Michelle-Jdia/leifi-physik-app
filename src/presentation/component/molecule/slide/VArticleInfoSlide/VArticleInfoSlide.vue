<script setup lang="ts">
import type { SolvedEnumId } from '@/data/type/api/solvedApi';
import type { FavoriteTypeKey } from '@/data/type/app/favorite';
import type { ContentType } from '@/data/type/helper/contentType';
import type { ClassValue } from 'clsx';
import type { ComputedRef } from 'vue';
import { IonItemOption } from '@ionic/vue';
import { bookmark, bookmarkOutline, chevronForwardOutline, document, trash } from 'ionicons/icons';
import { computed } from 'vue';
import { isFavoriteType } from '@/data/type/app/favorite';
import { cn } from '@/presentation/helper/style';
import VBadge from '@/atom/badge/VBadge/VBadge.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VLearnStatusPopover from '@/atom/popover/VLearnStatusPopover/VLearnStatusPopover.vue';
import VSlide from '@/atom/slide/VSlide/VSlide.vue';
import VTransition from '@/atom/transition/VTransition/VTransition.vue';

export interface Props {
    id: string;
    learnStatus?: SolvedEnumId | string;
    type:
        | ContentType.ISSUE
        | ContentType.TASK
        | ContentType.DOWNLOAD
        | ContentType.LINK
        | ContentType.TOPIC;
    branch: {
        title: string;
        color: string;
    };
    title: string;
    subtitle: string;
    hasNote?: boolean;
    isLearnMode?: boolean;
    isMyCollection?: boolean;
    isFavorite?: boolean;
    isTopic?: boolean;
    modifier?: ClassValue;
}

const props = defineProps<Props>();

interface onFavoriteOutput {
    id: string;
    type: FavoriteTypeKey;
}

const emit = defineEmits<{
    (e: 'onDelete', id: string): void;
    (e: 'onEditNote', id: string): void;
    (e: 'onFavorite', { id, type }: onFavoriteOutput, closeSlide: () => void): void;
    (
        e: 'onLearnStatusClick',
        { id, learnStatusId }: { id: string; learnStatusId: SolvedEnumId | string },
    ): void;
}>();

const favoriteIcon: ComputedRef<string> = computed(() => {
    return props.isFavorite ? bookmark : bookmarkOutline;
});

function emitSelectedLearnStatus(learnStatusId: SolvedEnumId | string): void {
    emit('onLearnStatusClick', {
        id: props.id,
        learnStatusId,
    });
}

function favorite(closeSlide: () => void): void {
    if (isFavoriteType(props.type)) {
        emit(
            'onFavorite',
            {
                id: props.id,
                type: props.type,
            },
            closeSlide,
        );
    }
}
</script>

<template>
    <v-slide
        :modifier="cn('article-info-slide', modifier, isTopic && 'v-bg-gray-20')"
        :slide-background="isTopic ? 'v-bg-gray-20' : ''"
    >
        <v-badge :color="branch.color" :modifier="{ '!v-bg-gray-50': isLearnMode }">
            {{ branch.title }}
        </v-badge>

        <v-badge v-if="isFavorite" modifier="v-ml-box v-bg-gray-50">Favorit</v-badge>

        <v-transition type="slide-right">
            <v-learn-status-popover
                v-if="isLearnMode"
                modifier="v-ml-box"
                :learn-status="learnStatus"
                @on-select="emitSelectedLearnStatus"
                @click.prevent
            />
        </v-transition>

        <div class="v-mt-box-sm">
            <h3 class="v-text-h4 md:v-text-h3">{{ title }}</h3>

            <span class="v-h5 v-mt-2 v-inline-block v-font-fflight v-text-ty-helper">
                {{ subtitle }}
            </span>
        </div>

        <div v-if="$slots.default" class="v-mt-box">
            <slot />
        </div>

        <template v-if="$slots.after" #after>
            <slot name="after" />
        </template>

        <template #end>
            <v-icon color="v-color-ty-headline" :icon="chevronForwardOutline" />
        </template>

        <template #actions="{ closeSlide }">
            <ion-item-option
                v-if="isMyCollection"
                data-testid="article-info-delete"
                class="v-w-[6rem] v-bg-danger v-text-center"
                @click="emit('onDelete', id)"
            >
                <v-icon slot="top" :icon="trash" />

                <span class="v-mt-box-md">LÃ¶schen</span>
            </ion-item-option>

            <ion-item-option
                v-if="isMyCollection"
                data-testid="article-info-create"
                class="v-w-[6rem] v-bg-warning v-text-center"
                @click="emit('onEditNote', id)"
            >
                <v-icon slot="top" :icon="document" />

                <span class="v-mt-box-md">
                    Notiz
                    <div v-if="hasNote">bearbeiten</div>

                    <div v-else>erstellen</div>
                </span>
            </ion-item-option>

            <!-- @TODO there should be a possibility to control from the VFavoriteWrapper.vue if the favorite functionality can be used here -->

            <ion-item-option
                v-if="isFavoriteType(type)"
                data-testid="article-info-favorite"
                class="v-w-[6rem] v-bg-pink v-text-center"
                @click="() => favorite(closeSlide)"
            >
                <v-icon slot="top" :icon="favoriteIcon" />

                <span class="v-mt-box-md">
                    Favorit
                    <div v-if="isFavorite">entfernen</div>

                    <div v-else>setzen</div>
                </span>
            </ion-item-option>
        </template>
    </v-slide>
</template>
