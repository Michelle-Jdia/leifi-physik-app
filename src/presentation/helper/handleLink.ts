import {
    isDownloadStringType,
    isIssueStringType,
    isTaskStringType,
} from '@/data/type/helper/contentType';
import { getContextualRouteNameByType } from '@/connection/helper/navigation';
import router from '@/connection/router';

function clickHandler(this: HTMLAnchorElement, event: MouseEvent) {
    const uuid = this.dataset.uuid;
    const type = this.dataset.type;

    if (
        !uuid ||
        !type ||
        !(isIssueStringType(type) || isDownloadStringType(type) || isTaskStringType(type))
    ) {
        return;
    }

    event.preventDefault();

    router.push({
        name: getContextualRouteNameByType(type),
        params: {
            id: uuid,
        },
    });

    this.removeEventListener('click', clickHandler);
}

export function handleLinksClick(containerElement: HTMLElement | null): void {
    if (!containerElement) {
        return;
    }

    const links = document.getElementsByTagName('a');

    Array.from(links).forEach((link: HTMLAnchorElement) => {
        link.addEventListener('click', clickHandler);
    });
}
