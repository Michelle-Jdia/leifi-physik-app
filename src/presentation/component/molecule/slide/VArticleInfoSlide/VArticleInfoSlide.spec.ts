import { render } from '@testing-library/vue';
import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';
import { testSnapshot } from '@/presentation/helper/test';
import { articleInfoSlideProps } from '@/molecule/slide/VArticleInfoSlide/VArticleInfoSlide';
import VArticleInfoSlide from '@/molecule/slide/VArticleInfoSlide/VArticleInfoSlide.vue';

describe('VArticleInfoSlide', () => {
    test('shows text that was passed to all props and shows 1 slide option', () => {
        const { getByText, getByTestId, queryByTestId } = render(VArticleInfoSlide, {
            props: {
                ...articleInfoSlideProps,
            },
        });

        getByText(articleInfoSlideProps.branch.title);
        getByText(articleInfoSlideProps.title);
        getByText(articleInfoSlideProps.subTitle);

        getByTestId('article-info-favorite');

        expect(queryByTestId('article-info-create')).toBeFalsy();
        expect(queryByTestId('article-info-delete')).toBeFalsy();
    });

    test('renders all slide options', async () => {
        const { getByTestId } = render(VArticleInfoSlide, {
            props: {
                ...articleInfoSlideProps,
                isLearnMode: true,
            },
        });

        getByTestId('article-info-favorite');
        getByTestId('article-info-create');
        getByTestId('article-info-delete');
    });

    test('emits correct events', async () => {
        const wrapper = mount(VArticleInfoSlide, {
            props: {
                ...articleInfoSlideProps,
                isLearnMode: true,
            },
        });

        await wrapper.find('[data-testid=article-info-create]').trigger('click');
        await wrapper.find('[data-testid=article-info-delete]').trigger('click');
        await wrapper.find('[data-testid=article-info-favorite]').trigger('click');

        const emitted = wrapper.emitted();

        expect(emitted.onCreate).toEqual([[articleInfoSlideProps.id]]);
        expect(emitted.onDelete).toEqual([[articleInfoSlideProps.id]]);
        // @todo figure our how to test emit callback
        // idea using JSON.stringify on both sides
        // expect(emitted.onFavorite).toBe([[articleInfoSlideProps.id, function closeSlide() {}]]);
    });

    testSnapshot(VArticleInfoSlide, {
        slots: {
            default: articleInfoSlideProps.branch,
        },
        props: {
            ...articleInfoSlideProps,
        },
    });
});
