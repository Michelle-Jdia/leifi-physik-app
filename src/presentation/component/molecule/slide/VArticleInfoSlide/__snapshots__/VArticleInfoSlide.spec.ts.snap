// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VArticleInfoSlide > snapshot 1`] = `
<v-slide-stub
  modifier="article-info-slide"
  slidebackground=""
>
  <v-badge-stub
    color="v-bg-yellow-80"
    modifier="[object Object]"
  >
    Mechanik
  </v-badge-stub>
  <!--v-if-->
  <v-transition-stub
    type="slide-right"
  >
    <!--v-if-->
  </v-transition-stub>
  <div
    class="v-mt-box-sm"
  >
    <h3
      class="v-text-h4 md:v-text-h3"
    >
      Title
    </h3>
    <span
      class="v-h5 v-mt-2 v-inline-block v-font-fflight v-text-ty-helper"
    >
      Subtitle
    </span>
  </div>
  <div
    class="v-mt-box"
  >
    
    <anonymous-stub />
    
  </div>
</v-slide-stub>
`;
